inference/core/workflows/core_steps/models/foundation/openai.py

```diff
@@ -0,0 +1,407 @@
+) -> List[Dict[str, str]]:
+    if remote_api_key is None:
+        raise ValueError(
+            "Step that involves GPT-4V prompting requires OpenAI API key which was not provided."
+        )
+    print("lmm_config", lmm_config)
+    return await execute_gpt_4v_requests(
+        image=image,
+        remote_api_key=remote_api_key,
+        prompt=prompt,
+        lmm_config=lmm_config,
+    )
+
+
+async def execute_gpt_4v_requests(
+    image: List[dict],
+    remote_api_key: str,
+    prompt: str,
+    lmm_config: LMMConfig,
+) -> List[Dict[str, str]]:
+    client = AsyncOpenAI(api_key=remote_api_key)
+    results = []
+    images_batches = list(
+        make_batches(
+            iterable=image,
+            batch_size=WORKFLOWS_REMOTE_EXECUTION_MAX_STEP_CONCURRENT_REQUESTS,
+        )
+    )
+    for image_batch in images_batches:
+        batch_coroutines = []
+        for image in image_batch:
+            coroutine = execute_gpt_4v_request(
+                client=client,
+                image=image,
+                prompt=prompt,
+                lmm_config=lmm_config,
+            )
+            batch_coroutines.append(coroutine)
+        batch_results = await asyncio.gather(*batch_coroutines)
+        results.extend(batch_results)
+    return results
+
+
+async def execute_gpt_4v_request(
+    client: AsyncOpenAI,
+    image: Dict[str, Any],
+    prompt: str,
+    lmm_config: LMMConfig,
+) -> Dict[str, str]:
+    loaded_image, _ = load_image(image)
+    image_metadata = {"width": loaded_image.shape[1], "height": loaded_image.shape[0]}
+    base64_image = base64.b64encode(encode_image_to_jpeg_bytes(loaded_image)).decode(
+        "ascii"
+    )
+    print("lmm_config", lmm_config)
+    response = await client.chat.completions.create(
+        model=lmm_config.gpt_model_version,
+        messages=[
+            {
+                "role": "user",
+                "content": [
+                    {"type": "text", "text": prompt},
+                    {
+                        "type": "image_url",
+                        "image_url": {
+                            "url": f"data:image/jpeg;base64,{base64_image}",
+                            "detail": lmm_config.gpt_image_detail,
+                        },
+                    },
+                ],
+            }
+        ],
+        max_tokens=lmm_config.max_tokens,
+    )
+    return {"content": response.choices[0].message.content, "image": image_metadata}
+
+
+def turn_raw_lmm_output_into_structured(
+    raw_output: List[Dict[str, Any]],
+    expected_output: Optional[Dict[str, str]],
+) -> List[dict]:
+    if expected_output is None:
+        return [{} for _ in range(len(raw_output))]
+    return [
+        try_parse_lmm_output_to_json(
+            output=r["content"],
+            expected_output=expected_output,
+        )
+        for r in raw_output
+    ]
+
+
+def try_parse_lmm_output_to_json(
+    output: str, expected_output: Dict[str, str]
+) -> Union[list, dict]:
```

Comment:
Just a thought and not for action, we support pytoon 3.8+; starting from python 3.9 using types like typing.Dict, typing.List is deprecated (https://peps.python.org/pep-0585/#implementation) - we seem to be using types from `typing` and generic types interchangeably , would be good to unify how we type